### ISSUE TITLE
Replay doesn't work with SAMPLING and OUTPUT_DB=false

### ISSUE THREAD

**Issue by stoyanr** [cloudfoundry-incubator/cf-abacus#560]
Currently, setting `SAMPLING` or setting `OUTPUT_DB=false` breaks the replay mechanism.
* With `SAMPLING`, input docs are correctly mapped to the single output doc, however if only some of them are not processed, they will not be replayed because the output doc already exists.
* With `OUTPUT_DB=false`, the output database doesn't exist at all, which breaks the replay logic.

We would very much like to rely on `SAMPLING` and `OUTPUT_DB=false` to address our issues with growing database size. At the same time, we would like to keep the replay mechanism. 

One possible solution would be to change the replay mechanism to not rely on documents in the output database. Instead, it could check:
* Special fields in the input documents that are only added after the document has been fully processed.
* Marker documents stored in a different collection / database, e.g. `REPLAY_DB` (`abacus-aggregator-replay-usage` or something similar).

**Comment by rajkiranrbala** [comment]
@stoyanr 
> With SAMPLING, input docs are correctly mapped to the single output doc, however, if only some of them are not processed, they will not be replayed because the output doc already exists.

This is not completely true. Sampling is applicable only for reducers. There need to be at least two docs emitted for every input doc of which the last doc(duplicate detection doc) must be unique and the other docs can be sampled.

> One possible solution would be to change the replay mechanism to not rely on documents in the output database. Instead, it could check:
>> Special fields in the input documents that are only added after the document has been fully processed. 

IMO this is a redundant operation. If we write the duplicate detection document, this would not be required. Also adding a DB write is an expensive process and we have to deal with conflicts.

>> Marker documents stored in a different collection / database, e.g. REPLAY_DB (abacus-aggregator-replay-usage or something similar).

I have to warn about using multiple databases. You might very likely run into the issue described here  https://github.com/cloudfoundry-incubator/cf-abacus/issues/469. Instead, we could have a flag that can be said as to what docs need to go into the output DB. In this way, we don't write all the documents to the output db and just the duplicate detection docs(Marker docs).

**Comment by stoyanr** [comment]
@rajkiranrbala Thanks for this information. We had a deeper look at the replay code. We found out that replay was using the "normal" output docs, not the duplicate detection docs. Since we want replay to work also with SAMPLING, we fixed it to use the duplicate detection docs, as you recommended.

Also, we found out that replay was using the document before it was "undbified", which worked correctly only with MongoDB. We fixed that as well and tested it on other databases.

The PR is #563.
